aicsimageio/readers/reader.py

```diff
@@ -0,0 +1,393 @@
+        """
+        pass
+
+    @classmethod
+    def assert_reader_supports_image(cls, image: types.ImageLike) -> bool:
+        """
+        Asserts that the provided image like object is supported by the current Reader.
+
+        Parameters
+        ----------
+        image: types.ImageLike
+            The filepath or array to validate as a supported type.
+
+        Returns
+        -------
+        supported: bool
+            Boolean indicated if the provided data is or is not supported by the
+            current Reader.
+
+        Raises
+        ------
+        TypeError: Invalid type provided to image parameter.
+        """
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def _assert_reader_supports_image(image: types.ImageLike) -> bool:
+        """
+        The per-Reader implementation of validating that an image is supported or not by
+        the Reader itself.
+        """
+        pass
+
+    @abstractmethod
+    def _read_delayed(self) -> da.Array:
+        """
+        The delayed dask array constructor for the image.
+
+        Returns
+        -------
+        dask_data: da.Array
+            The fully constructed delayed dask array.
+
+            It is additionally recommended to closely monitor how dask array chunks are
+            managed.
+
+        Notes
+        -----
+        This function will only be routed to under specific situations:
+            * use_dask(True) has been set OR
+            * the user has so far only used "dask_data" API calls
+            (get_image_dask_data / dask_data)
+        """
+        pass
+
+    @abstractmethod
+    def _read_immediate(self) -> np.ndarray:
+        """
+        The immediate numpy array constructor for the image.
+
+        Returns
+        -------
+        data: np.ndarray
+            The fully read numpy array.
+
+        Notes
+        -----
+        The implementer should assume that this function is the primary route for their
+        image reading. See `_read_delayed` for more details.
+        """
+        pass
+
+    @property
+    def dask_data(self) -> da.Array:
+        """
+        Returns
+        -------
+        dask_data: da.Array
+            The image as a dask array with the native dimension ordering.
+        """
+        pass
+
+    @property
+    def data(self) -> np.ndarray:
+        """
+        Returns
+        -------
+        data: np.ndarray
+            The image as a numpy array with native dimension ordering.
+        """
+        pass
+
+    @property
+    def dtype(self) -> np.dtype:
+        """
+        Returns
+        -------
+        dtype: np.dtype
+            Data-type of the image array's elements.
+        """
+        pass
+
+    @property
+    def shape(self) -> Tuple[int]:
+        """
+        Returns
+        -------
+        shape: Tuple[int]
+            Tuple of the image array's dimensions.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def dims(self) -> Dimensions:
```

Comment:
Dims and related shape functions: 
What's the desired behavior when the images in the file have inconsistent shape?

Reply:
That is a big big problem for the aicsimage class because it is assuming a 6d numpy array under the hood.  That implies uniformity.    Maybe we need to separate out "S" and deal with a "list" of 5d arrays which could each have a different shape.

Reply:
Even ome-xml has the same issue, I believe. Each <image> can have totally separate pixel counts.